packages/state-transition/src/cache/shufflingCache.ts

```diff
@@ -0,0 +1,237 @@
+import {Epoch, RootHex} from "@lodestar/types";
+import {GaugeExtra, MapDef, NoLabels, pruneSetToMax} from "@lodestar/utils";
+import {EpochShuffling, computeEpochShuffling} from "../util/index.js";
+import {BeaconStateAllForks} from "./types.js";
+
+/**
+ * With default chain option of maxSkipSlots = 32, there should be no shuffling promise. If that happens a lot, it could blow up Lodestar,
+ * with MAX_EPOCHS = 4, only allow 2 promise at a time. Note that regen already bounds number of concurrent requests at 1 already.
+ */
+export const SHUFFLING_CACHE_MAX_PROMISES = 2;
+
+/**
+ * Same value to CheckpointBalancesCache, with the assumption that we don't have to use it for old epochs. In the worse case:
+ * - when loading state bytes from disk, we need to compute shuffling for all epochs (~1s as of Sep 2023)
+ * - don't have shuffling to verify attestations, need to do 1 epoch transition to add shuffling to this cache. This never happens
+ * with default chain option of maxSkipSlots = 32
+ **/
+export const SHUFFLING_CACHE_MAX_EPOCHS = 4;
+
+export enum ShufflingCacheCaller {
```

Comment:
If we don't actually need the caller, then we can move the implementation to beacon node like we talked about.